### PR TITLE
TO BE SPLITTED: Refactor/organize folder structure

### PR DESCRIPTION
**Pull request by georgehristov**
- Move DB specific code to designated namespaces
- Introduce Connection class registry based on driverType
- Utilize object inheritance for setting up connection specific settings and behavior
- Improve variable naming consistency
- Change variable and method names based on semantics
- Keep full BC

### CODE REVIEW

File: src/Connection.php [hunk 1]
```diff
@@ -9,17 +9,39 @@ class Connection
...
-    protected $expression_class = Expression::class;
+    protected $expressionClass = Expression::class;
 
     /** @var Connection|\PDO Connection or PDO object */
+    protected $handler;
+
+    /**
+     * @deprecated use $handler instead
+     */
     protected $connection;
```
Comment (mvorisek) [inline src/Connection.php:50]: This is protected, there does not need to be BC. We are not Google.

File: src/Connection.php [hunk 2]
```diff
@@ -9,17 +9,39 @@ class Connection
...
 
     /** @var int Current depth of transaction */
-    public $transaction_depth = 0;
+    public $transactionDepth = 0;
+
+    /**
+     * @deprecated use $transactionDepth instead
+     */
+    public $transaction_depth;
```
Comment (mvorisek) [inline src/Connection.php]: If used somewhere, PHP will emit a notice that you try to use an unitialized property.

File: src/Connection.php [hunk 3]
```diff
@@ -29,22 +51,38 @@ class Connection
...
     {
         if (!is_array($properties)) {
             throw new Exception([
-                'Invalid properties for "new Connection()". Did you mean to call Connection::connect()?',
+                'Invalid properties for "new Connection()". Did you mean to call Connection::create()?',
                 'properties' => $properties,
             ]);
         }
 
+        $this->driverType = static::DEFAULT_DRIVER_TYPE;
```
Comment (mvorisek) [inline src/Connection.php]: If default type should be null, then there should be not const like that.

File: src/Connection.php [hunk 4]
```diff
@@ -97,7 +135,15 @@ public static function normalizeDSN($dsn, $user = null, $pass = null)
     }
 
     /**
-     * Connect database.
+     * @deprecated use Connection::create instead
+     */
+    public static function connect($dsn, $user = null, $password = null, $args = [])
```
Comment (mvorisek) [inline src/Connection.php:143]: With each meaningful deprecation I would you like to see when it should be removed. We need to reduce to the code as much as possible.

File: src/Connection.php [hunk 5]
```diff
@@ -106,125 +152,98 @@ public static function normalizeDSN($dsn, $user = null, $pass = null)
      *
      * @return Connection
      */
-    public static function connect($dsn, $user = null, $password = null, $args = [])
+    public static function create($dsn, $user = null, $password = null, $args = [])
```
Comment (mvorisek) [inline src/Connection.php:158]: When you create a Pdo, does it try to connect immediatelly?

File: src/Connection.php [hunk 6]
```diff
@@ -106,125 +152,98 @@ public static function normalizeDSN($dsn, $user = null, $pass = null)
...
+     * MySQL\Connection::register()
+     *
+     * CustomDriver\Connection must be descendant of Connection class.
+     *
+     * @param string $driverType
+     * @param string $connectionClass
+     */
+    public static function register($driverType = null, $connectionClass = null)
+    {
+        if (!$connectionClass && is_a($driverType, self::class, true)) {
```
Comment (mvorisek) [inline src/Connection.php]: This class can be abstract if needed. Do we need direct instance of Connection anywhere?

File: src/Connection.php [hunk 7]
```diff
@@ -106,125 +152,98 @@ public static function normalizeDSN($dsn, $user = null, $pass = null)
...
+    }
+
+    /**
+     * Resolves the connection class to use based on driver type.
+     *
+     * @param string $driverType
+     *
+     * @return string
+     */
+    public static function resolve($driverType)
```
Comment (mvorisek) [inline src/Connection.php]: Too generic name

File: src/Connection.php [hunk 8]
```diff
@@ -236,48 +255,56 @@ protected static function getPDO(array $dsn)
...
+        return $this->handler();
+    }
+
+    /**
+     * Returns the connection handler.
      *
      * @return Connection|\PDO
      */
-    public function connection()
+    public function handler()
```
Comment (mvorisek) [inline src/Connection.php]: Driver seems more industry common to me here.

Is it desireable to allow - use case? - `\Connection` here at all?

File: src/Counter/Connection.php [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+namespace atk4\dsql\Counter;
+
+use atk4\dsql\Expression;
+use atk4\dsql\ProxyConnection;
+use atk4\dsql\Query;
+
+class Connection extends ProxyConnection
+{
+    const DEFAULT_DRIVER_TYPE = 'counter';
```
Comment (mvorisek) [inline src/Counter/Connection.php]: Why not simply DRIVER_TYPE ?

File: src/Expression.php [hunk 10]
```diff
@@ -41,7 +41,7 @@ class Expression implements \ArrayAccess, \IteratorAggregate
      *
      * @var string
      */
-    protected $escape_char = '"';
+    protected $quotedIdentifier = '"';
```
Comment (mvorisek) [inline src/Expression.php]: Identifier? `escapeChar` seems better to me, putting here something more than a char probably is wrong.

File: src/Connection.php [hunk 11]
```diff
@@ -236,48 +255,56 @@ protected static function getPDO(array $dsn)
...
+        return $this->handler ?? $this;
     }
 
     /**
      * Execute Expression by using this connection.
      *
      * @return \PDOStatement
      */
-    public function execute(Expression $expr)
+    public function execute(Expression $expression)
```
Comment (mvorisek) [inline src/Connection.php]: On ther places there is `$expr`, usually I do not like shortening, but it makes sense here (keep the original).

**Review by mvorisek — CHANGES_REQUESTED** [review]

File: src/Connection.php (hunk 1 above)
Comment (georgehristov) [reply src/Connection.php:50]: The problem is `DIContainerTrait` lets setting protected properties from external locations.
Some of the legacy projects might use `['connection' => Whatever]` so I decided to keep it BC for time being.

File: src/Connection.php (hunk 3 above)
Comment (georgehristov) [reply src/Connection.php]: `static::DEFAULT_DRIVER_TYPE` is defined in each child class

File: src/Connection.php (hunk 5 above)
Comment (georgehristov) [reply src/Connection.php:158]: The semantics is `Connection::create`. So you create a connection and it connects

File: src/Expression.php (hunk 10 above)
Comment (georgehristov) [reply src/Expression.php]: This is not an escape string but quoted identifier.
https://dba.stackexchange.com/questions/23129/benefits-of-using-backtick-in-mysql-queries

File: src/Counter/Connection.php (hunk 9 above)
Comment (georgehristov) [reply src/Counter/Connection.php]: The actual driver type is in `$driverType` property.
`DEFAULT_DRIVER_TYPE` is the default one that will be used if no other driverType is set
I spend a few hours figuring out how to name it. Naming it `DRIVER_TYPE` does not hint to what the actual use is and what is the difference with the property `$driverType`

File: src/Connection.php (hunk 1 above)
Comment (mvorisek) [reply src/Connection.php:50]: Wdyt to update `DIContainerTrait` which throw an exception (with hint) when deprecated property is set with it?

Or even a new core trait which will accept list of deprecated properties/methods and throw an exception when any of them are accessed (using magic methods/already existing trait).

File: src/Connection.php (hunk 5 above)
Comment (mvorisek) [reply src/Connection.php:158]: I am okwith it, only a question if the rename is worth enought.

File: src/Counter/Connection.php (hunk 9 above)
Comment (mvorisek) [reply src/Counter/Connection.php]: What is the problem with setting `$driverType` to the default value?

File: src/Connection.php (hunk 8 above)
Comment (georgehristov) [reply src/Connection.php]: Using `handler` here also needed some thought. Handler can be a `Connection` and `PDO` or some other class so it has use cases and does not confuse the dev having `Connection::connection`
Handler handles the connection

File: src/Expression.php (hunk 10 above)
Comment (mvorisek) [reply src/Expression.php]: Is this guarantted to be never used with values and only with names?

File: src/Connection.php (hunk 8 above)
Comment (mvorisek) [reply src/Connection.php]: I got confused myself in the past... Ok to rename to `driver`?

File: src/Connection.php (hunk 1 above)
Comment (georgehristov) [reply src/Connection.php:50]: Yes, good idea!

File: src/Counter/Connection.php (hunk 9 above)
Comment (georgehristov) [reply src/Counter/Connection.php]: My idea was to have a way of retrieving the default driver type for a connection class without having to instantiate it. (e.g `MySQL\Connection::DEFAULT_DRIVER_TYPE`)
This is used in `Connection::register` where the connection class is registered for its default driver type if no other set specifically.

Comment (mvorisek) [reply src/Counter/Connection.php]: I am not sure if that can be done using reflection, but probably yes :D

Comment (mvorisek) [reply src/Counter/Connection.php]: https://www.php.net/manual/en/reflectionclass.getdefaultproperties.php

File: src/Connection.php (hunk 11 above)
Comment (georgehristov) [reply src/Connection.php]: Reverted

File: src/Counter/Connection.php (hunk 9 above)
Comment (georgehristov) [reply src/Counter/Connection.php]: Used your suggestion and got rid of the constant altogether

File: src/Connection.php (hunk 8 above)
Comment (georgehristov) [reply src/Connection.php]: Renamed to driver now if it is more understandable

File: src/Connection.php (hunk 7 above)
Comment (georgehristov) [reply src/Connection.php]: Renamed to `resolveDriver`

File: src/Expression.php (hunk 10 above)
Comment (georgehristov) [reply src/Expression.php]: This string is used to wrap (quote) literal strings to be used in the query like in MySQL
```
CREATE TABLE `my table` (id INT);
```

Comment (mvorisek) [reply src/Expression.php]: If, then `quoteIdentifier`.

But Microsoft https://docs.microsoft.com/en-us/sql/t-sql/statements/set-quoted-identifier-transact-sql uses this term for string values not string identifiers..

File: src/Counter/Connection.php (hunk 9 above)
Comment (mvorisek) [reply src/Counter/Connection.php]: 👍

File: src/Connection.php (hunk 7 above)
Comment (mvorisek) [reply src/Connection.php]: 👍 wdyt about the same for registry, register() method..?

Comment (georgehristov) [reply src/Connection.php]: I believe keeping only `register` is more clear and understandable.
`CustomConnection::register();` or
`CustomConnection::register('forSomeDriverType');` or 
`Connection::register('someDriverType', CustomConnection::class)`

Comment (mvorisek) [reply src/Connection.php]: I meant `registerDriver`, dyt it is less clear?

Comment (georgehristov) [reply src/Connection.php]: Yes, I understood about `registerDriver`. I think now is more clear but let me think a bit more about it.

Comment (mvorisek) [reply src/Connection.php]: Or even `registerDriverClass`.

Comment (georgehristov) [reply src/Connection.php]: Now is reformatted and more understandable I believe

File: src/Connection_Oracle.php [hunk 12]
```diff
@@ -3,8 +3,8 @@
 namespace atk4\dsql;
 
 /**
- * @deprecated use Oracle\Connection instead
+ * @deprecated use Oracle\Driver instead
  */
-class Connection_Oracle extends Oracle\Connection
+class Connection_Oracle extends Oracle\Driver
```
Comment (mvorisek) [inline src/Connection_Oracle.php:8]: Is it better? I meant the to only refactor the handler to driver, driver term is usually used for `\Pdo` driver.

Comment (georgehristov) [reply src/Connection_Oracle.php:8]: The thing is that an object of Connection class can also be a driver.
So the DB specific implementation of Connection we call Driver and we register it

Comment (mvorisek) [reply src/Connection_Oracle.php:8]: I see... But it extend `Connection`, so using 2 terms for the same seems not good to me :)

Comment (georgehristov) [reply src/Connection_Oracle.php:8]: The mice are playing but let's see what the big cats @romaninsh and @DarkSide666 will say about all this :)

File: src/MySQL/Driver.php [hunk 13]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace atk4\dsql\MySQL;
+
+use atk4\dsql\Connection;
+
+class Driver extends Connection
+{
+    public $driverType = 'mysql';
```
Comment (mvorisek) [inline src/MySQL/Driver.php:9]: What is the actual purpose of driverType, when we can ceck if the class is instanceof? Should be probably removed.

Comment (georgehristov) [reply src/MySQL/Driver.php:9]: Refactored a bit more https://github.com/atk4/dsql/tree/refactor/organize-folder-structure-driver

WDYT?

**Comment by DarkSide666** [comment]
Huge amount of backward incompatible things :(

**Comment by georgehristov** [comment]
> 
> 
> Huge amount of backward incompatible things :(

This is not the idea. In my opinion should be BC but needs to be tested on actual projects

**Comment by mvorisek** [comment]
@georgehristov can you update this PR to start at current develop without any folder renames? (we can check and decise with folder structure rearrange later)

**Comment by georgehristov** [comment]
Folder structure is major part of the refactoring. And all is intended / tested to be BC.
I have similar folder restructure lined up for the other repos as well.

**Comment by mvorisek** [comment]
@georgehristov There were many changes, this is primary to do a review, see real changes and then we can reupdate the structure very easily.

**Comment by mvorisek** [comment]
@georgehristov PRs like this are bad. They can not be checked nor merge if not done so immediatelly.

Is it ok with you to split this PR at least into two - one for pure folder/NS refactor and another for changes?

File: src/MySQL/Driver.php (hunk 13 above)
Comment (mvorisek) [reply src/MySQL/Driver.php:9]: @georgehristov 

1. huge 👍 for splitting this PR. In the future, it is really better to isolate larger changes into more isolated oned, good job!
2. is dsql/218 the last part?
3. here are some comments to which seems to be not addressed

at least:
- about the `driverType` - is seems fully redundant for and I think we should test `instanceof Engine/Connection` to the right type (also much more refactorable/usable)
- in the new PR you use still `driver` word
- is it guaranteed that `create` does only create the underlaying driver and does not `connect`? (as with connect - I count with DB ping/issues, and with create - I expect no connection error/exception)

**Pull request closed without merging**